libs/net/ches/ChesService.cs

```diff
@@ -0,0 +1,369 @@
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web;
+
+namespace TNO.Ches
+{
+    /// <summary>
+    /// ChesService class, provides a service for integration with Ches API services.
+    /// </summary>
+    public class ChesService : IChesService
+    {
+        #region Variables
+        private readonly ClaimsPrincipal _user;
+        private TokenModel _token = null;
+        private readonly JwtSecurityTokenHandler _tokenHandler;
+        private readonly ILogger<IChesService> _logger;
+        #endregion
+
+        #region Properties
+        protected OpenIdConnectRequestClient Client { get; }
+        public ChesOptions Options { get; }
+        #endregion
+
+        #region Constructors
+        /// <summary>
+        /// Creates a new instance of a ChesService, initializes with specified arguments.
+        /// </summary>
+        /// <param name="options"></param>
+        /// <param name="user"></param>
+        /// <param name="client"></param>
+        /// <param name="tokenHandler"></param>
+        /// <param name="logger"></param>
+        public ChesService(IOptions<ChesOptions> options, ClaimsPrincipal user, OpenIdConnectRequestClient client, JwtSecurityTokenHandler tokenHandler, ILogger<IChesService> logger)
+        {
+            this.Options = options.Value;
+            _user = user;
+            this.Client = client;
+            _tokenHandler = tokenHandler;
+            _logger = logger;
+        }
+        #endregion
+
+        #region Methods
+        /// <summary>
+        /// Generates the full URL including the host.
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <param name="outputFormat"></param>
+        /// <returns></returns>
+        private string GenerateUrl(string endpoint)
+        {
+            return $"{this.Options.HostUri}{endpoint}";
+        }
+
+        /// <summary>
+        /// Ensure we have an active access token.
+        /// Make an HTTP request if one is needed.
+        /// </summary>
+        /// <returns></returns>
+        private async Task RefreshAccessTokenAsync()
+        {
+            // Check if token has expired.  If it has refresh it.
+            if (_token == null || String.IsNullOrWhiteSpace(_token.AccessToken) || _tokenHandler.ReadJwtToken(_token.AccessToken).ValidTo <= DateTime.UtcNow)
+            {
+                _token = await GetTokenAsync();
+            }
+        }
+
+        /// <summary>
+        /// Send a request to the specified endpoint.
+        /// </summary>
+        /// <typeparam name="TR"></typeparam>
+        /// <param name="endpoint"></param>
+        /// <param name="method"></param>
+        /// <returns></returns>
+        private async Task<string> SendAsync(string endpoint, HttpMethod method)
+        {
+            await RefreshAccessTokenAsync();
```

Comment:
Shouldn't be required because `IOpenIdConnectRequestClient` should do this automatically when making requests.